server/src/main/java/org/elasticsearch/ExceptionsHelper.java

```diff
@@ -278,6 +278,20 @@ public static ShardOperationFailedException[] groupBy(ShardOperationFailedExcept
         return uniqueFailures.toArray(new ShardOperationFailedException[0]);
     }
 
+    /**
+     * @param t Throwable to inspect
+     * @return true if the Throwable is an instance of an Exception that indicates
+     *         that either a Node or shard is unavailable/disconnected.
+     */
+    public static boolean isNodeOrShardUnavailableTypeException(Throwable t) {
+        return (t instanceof org.elasticsearch.action.NoShardAvailableActionException
+            || t instanceof org.elasticsearch.action.UnavailableShardsException
+            || t instanceof org.elasticsearch.node.NodeClosedException
+            || t instanceof org.elasticsearch.transport.NodeDisconnectedException
+            || t instanceof org.elasticsearch.discovery.MasterNotDiscoveredException
+            || t instanceof org.elasticsearch.transport.NodeNotConnectedException);
```

Comment:
ClusterBlockException could be one of them too. It may deserve some double checking at least.